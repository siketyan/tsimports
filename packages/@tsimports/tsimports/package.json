{
  "name": "@tsimports/tsimports",
  "version": "0.1.3",
  "description": "A fast and opinionated imports organizer for ECMAScript and TypeScript.",
  "author": "Naoki Ikeguchi <me@s6n.jp>",
  "license": "MIT",
  "homepage": "https://github.com/siketyan/tsimports",
  "bugs": "https://github.com/siketyan/tsimports/issues",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/siketyan/tsimports.git",
    "directory": "packages/tsimports"
  },
  "publishConfig": {
    "provenance": true
  },
  "engines": {
    "node": ">=18"
  },
  "files": ["bin/tsimports", "scripts/postinstall.js", "README.md", "LICENCE.md"],
  "bin": {
    "tsimports": "bin/tsimports"
  },
  "optionalDependencies": {
    "@tsimports/cli-darwin-arm64": "0.1.3",
    "@tsimports/cli-darwin-x64": "0.1.3",
    "@tsimports/cli-linux-arm64": "0.1.3",
    "@tsimports/cli-linux-x64": "0.1.3"
  },
  "scripts": {
    "postinstall": "node scripts/postinstall.js"
  }
}
