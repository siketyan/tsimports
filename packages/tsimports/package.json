{
  "name": "tsimports",
  "version": "0.1.2",
  "description": "A fast and opinionated imports organizer for ECMAScript and TypeScript.",
  "author": "Naoki Ikeguchi <me@s6n.jp>",
  "license": "MIT",
  "homepage": "https://github.com/siketyan/tsimports",
  "bugs": "https://github.com/siketyan/tsimports/issues",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/siketyan/tsimports.git",
    "directory": "packages/tsimports"
  },
  "publishConfig": {
    "provenance": true
  },
  "engines": {
    "node": ">=18"
  },
  "files": ["bin/tsimports", "scripts/postinstall.js", "README.md", "LICENCE.md"],
  "bin": {
    "tsimports": "bin/tsimports"
  },
  "optionalDependencies": {
    "@tsimports/cli-darwin-arm64": "0.1.2",
    "@tsimports/cli-darwin-x64": "0.1.2",
    "@tsimports/cli-linux-arm64": "0.1.2",
    "@tsimports/cli-linux-x64": "0.1.2"
  },
  "scripts": {
    "postinstall": "node scripts/postinstall.js"
  }
}
