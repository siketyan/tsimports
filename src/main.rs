use std::fs::{read_to_string, write};
use std::io::{stdin, Read};
use std::path::PathBuf;

use anyhow::{anyhow, Result};
use biome_console::{markup, ColorMode, Console, EnvConsole, LogLevel};
use biome_diagnostics::PrintDiagnostic;
use biome_js_syntax::JsFileSource;
use clap::{Parser, ValueEnum};
use tsimports::{tsimports, Error};

#[derive(Copy, Clone, Debug, Default, ValueEnum)]
enum Language {
    #[default]
    JS,
    JSX,
    TS,
    TSX,
}

#[derive(Debug, Parser)]
#[command(about, version)]
struct Args {
    /// Path of the file to organize imports. Defaults to the standard input.
    path: Option<PathBuf>,

    /// Specify the language to parse the input as. Inferred by the file extension by default.
    #[clap(short, long)]
    language: Option<Language>,

    /// Write the formatted result into the file directly, without printing to the standard output.
    #[clap(short, long)]
    write: bool,
}

fn main() {
    let mut console = EnvConsole::new(ColorMode::Auto);
    if let Err(e) = run(&mut console) {
        console.println(
            LogLevel::Error,
            markup! { <Error>"\u{2716} "{e.to_string()}</Error> },
        )
    }
}

fn run(console: &mut impl Console) -> Result<()> {
    let args = Args::parse();

    let mut source = JsFileSource::js_module();
    let input = match &args.path {
        Some(path) => {
            if let Some(ext) = path.extension().and_then(|ext| ext.to_str()) {
                match ext {
                    "jsx" => source = JsFileSource::jsx(),
                    "ts" => source = JsFileSource::ts(),
                    "tsx" => source = JsFileSource::tsx(),
                    _ => {}
                }
            }

            read_to_string(path)?
        }
        None => {
            let mut buf = String::new();
            stdin().read_to_string(&mut buf)?;
            buf
        }
    };

    if let Some(lang) = args.language {
        use Language::*;

        source = match lang {
            JS => JsFileSource::js_module(),
            JSX => JsFileSource::jsx(),
            TS => JsFileSource::ts(),
            TSX => JsFileSource::tsx(),
        }
    } else {
        console.println(LogLevel::Error, markup! {
            <Warn>
                "\u{26a0} Input language is not specified, assuming as an ECMAScript module. Use "
                <Emphasis>"--language <js|jsx|ts|tsx>"</Emphasis>
                " option to override."
            </Warn>
        });
    };

    let output = match tsimports(input.as_str(), source) {
        Ok(o) => o,
        Err(e) => match e {
            Error::Parser(diags) => {
                for diag in diags {
                    console.println(
                        LogLevel::Error,
                        markup! { {PrintDiagnostic::verbose(&diag)} },
                    );
                }

                std::process::exit(1);
            }
            _ => return Err(anyhow!(e)),
        },
    };

    if args.write {
        let Some(path) = args.path else {
            return Err(anyhow!(
                "Can't write the result as the input was from the standard input."
            ));
        };

        write(path, output)?;
    } else {
        print!("{}", output);
    }

    Ok(())
}
