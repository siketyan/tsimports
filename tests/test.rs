use biome_js_syntax::JsFileSource;
use insta::assert_snapshot;
use tsimports::tsimports;

#[test]
fn snapshot_tests() {
    assert_snapshot!(tsimports(
        "\
import fs from 'fs'
import path from 'path'
import _ from 'lodash'
import chalk from 'chalk'
import foo from 'src/foo'
import foo from '../foo'
import qux from '../../foo/qux'
import bar from './bar'
import baz from './bar/baz'
import main from './'
import log = console.log
import type { Foo, Bar } from 'foo'
import userEvent from '@testing-library/user-event'
import { foo, type bar } from '@/my/module'
import * as R from 'remeda'
import * as utils from '~/utils'
import {} from '#import'
",
        JsFileSource::ts(),
    )
    .unwrap());
}
