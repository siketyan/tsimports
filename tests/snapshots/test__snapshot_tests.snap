---
source: tests/test.rs
expression: "tsimports(\"\\\nimport fs from 'fs'\nimport path from 'path'\nimport _ from 'lodash'\nimport chalk from 'chalk'\nimport foo from 'src/foo'\nimport foo from '../foo'\nimport qux from '../../foo/qux'\nimport bar from './bar'\nimport baz from './bar/baz'\nimport main from './'\nimport log = console.log\nimport type { Foo, Bar } from 'foo'\nimport userEvent from '@testing-library/user-event'\nimport { foo, type bar } from '@/my/module'\nimport * as R from 'remeda'\nimport * as utils from '~/utils'\nimport {} from '#import'\n\",\nJsFileSource::ts(),).unwrap()"
---
import fs from 'fs';
import path from 'path';

import userEvent from '@testing-library/user-event';
import chalk from 'chalk';
import type { Bar, Foo } from 'foo';
import _ from 'lodash';
import * as R from 'remeda';
import foo from 'src/foo';

import {  } from '#import';
import { type bar, foo } from '@/my/module';
import * as utils from '~/utils';

import qux from '../../foo/qux';
import foo from '../foo';

import bar from './bar';
import baz from './bar/baz';

import main from './';

import log = console.log
